### DISCUSSION TITLE
Re-Training SentenceRecognizer

### DISCUSSION THREAD

**Discussion by roelvanderburg** [explosion/spaCy#6926] [Help: Coding & Implementations] [answered]
I'm trying to retrain the SentenceRecognizer for Dutch data and have a small dataset of **25853** sentences. 
With the code below I'm splitting the sentences into docs and into a train and dev set. 

```
import spacy
from spacy.tokens import Doc
from spacy.tokens import DocBin

nlp = spacy.blank("nl")
db = DocBin()

X_sentence_train = all_sentence_list[1500:]
X_sentence_dev   = all_sentence_list[:1500]

i = 0
while i < len(X_sentence_train):

    # some realistic paragraph sizes
    para_size = random.randint(3, 15)
    # create a Doc for each sentence
    sent_docs = [nlp(sent) for sent in X_sentence_train[i : i + para_size]]
    # merge the docs together (adding a single space between docs)
    doc = Doc.from_docs(sent_docs, ensure_whitespace=True)
    # add to the DocBin
    db.add(doc)
    i += para_size
    
db.to_disk("train.spacy")
```

Now after training 

`spacy train config.cfg --paths.train train.spacy --paths.dev dev.spacy -o output_dir
`

I find the model still has some issues with certain abbreviations and lists.  Is there a way I could improve the performance?

I was thinking about increasing the windows size of the token2vec and/or changing optimizer? Any other suggestions?
Or is there a way to add a dictionary of abbreviations words

[paths]
train = null
dev = null
vectors = null
init_tok2vec = null

[system]
gpu_allocator = null
seed = 0

[nlp]
lang = "nl"
pipeline = ["tok2vec","senter"]
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null

[components]

[components.senter]
factory = "senter"

[components.senter.model]
@architectures = "spacy.Tagger.v1"
nO = null

[components.senter.model.tok2vec]
@architectures = "spacy.HashEmbedCNN.v1"
pretrained_vectors = null
width = 12
depth = 1
embed_size = 2000
window_size = 1
maxout_pieces = 2
subword_features = true

[components.tok2vec]
factory = "tok2vec"

[components.tok2vec.model]
@architectures = "spacy.Tok2Vec.v1"

[components.tok2vec.model.embed]
@architectures = "spacy.MultiHashEmbed.v1"
width = ${components.tok2vec.model.encode.width}
attrs = ["NORM","PREFIX","SUFFIX","SHAPE"]
rows = [5000,2500,2500,2500]
include_static_vectors = false

[components.tok2vec.model.encode]
@architectures = "spacy.MaxoutWindowEncoder.v1"
width = 96
depth = 4
window_size = 1
maxout_pieces = 3

[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
max_length = 2000
gold_preproc = false
limit = 0
augmenter = null

[training]
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
accumulate_gradient = 1
patience = 1600
max_epochs = 0
max_steps = 20000
eval_frequency = 200
frozen_components = []
before_to_disk = null

[training.batcher]
@batchers = "spacy.batch_by_words.v1"
discard_oversize = false
tolerance = 0.2
get_length = null

[training.batcher.size]
@schedules = "compounding.v1"
start = 100
stop = 1000
compound = 1.001
t = 0.0

[training.logger]
@loggers = "spacy.ConsoleLogger.v1"
progress_bar = false

[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = false
eps = 0.00000001
learn_rate = 0.001

[training.score_weights]
sents_f = 1.0
sents_p = 0.0
sents_r = 0.0

[pretraining]

[initialize]
vectors = null
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
lookups = null

[initialize.components]

**Comment by adrianeboyd** [comment, marked as answer]
Abbreviations are typically handled as tokenizer exceptions so that you end up with one token like `etc.` instead of `etc .`, which will help stop them from being mistaken as sentence boundaries. See: https://spacy.io/usage/linguistic-features#special-cases

```python
nlp.tokenizer.add_special_case("etc.", [{"ORTH": "etc."}])
```

When you save the model with `nlp.to_disk()` the exceptions are saved with the tokenizer settings.

In your config, you'd want to add a registered callback function to modify the tokenizer in `[initialize.before_init]` like this: https://spacy.io/usage/training#custom-tokenizer

----------

You can drop the `tok2vec` component from your pipeline because it's unused. (We should modify the quickstart/init to skip it when there are no listeners.)

The default `senter` component has its own internal `tok2vec` and its parameters are intentionally as small as possible for reasonably good performance to make it very fast and small on disk. You would very likely notice improvements in increasing `width`, `depth`, and `embed_size`, where `width` and `depth` are probably the best place to start.

```ini
[components.senter.model.tok2vec]
@architectures = "spacy.HashEmbedCNN.v1"
pretrained_vectors = null
width = 12
depth = 1
embed_size = 2000
window_size = 1
maxout_pieces = 2
subword_features = true
```

----------

You could also try adding `SPACY` as a feature to the senter tok2vec, which is whether the token has a following space. This can help with things like ASCII quotes, where the following whitespace can tell you which sentence it belongs to. To do this you'd have to switch to a more detailed `tok2vec` config that looks like the default `tok2vec` from above rather than using `spacy.HashEmbedCNN.v1`, so that you can specify the `attrs` rather than just the option `subword_features`. You need to add `SPACY` to `attrs` and add a size for the number of rows in the hash table to `rows`:

```ini
[components.senter.model.tok2vec]
@architectures = "spacy.Tok2Vec.v2"

[components.senter.model.tok2vec.embed]
@architectures = "spacy.MultiHashEmbed.v1"
width = ${components.senter.model.tok2vec.encode.width}
attrs = ["NORM","PREFIX","SUFFIX","SHAPE","SPACY"]
rows = [5000,2500,2500,2500,20]
include_static_vectors = false

[components.senter.model.tok2vec.encode]
@architectures = "spacy.MaxoutWindowEncoder.v2"
width = 12
depth = 1
window_size = 1
maxout_pieces = 2
```

As a note: the multi hash embed isn't really designed for boolean features like `SPACY`, so the number of rows looks a bit weird for a boolean value. The problem is that each value is hashed into four rows in the table and if the size is too small, you can easily end up with collisions where you can't distinguish the two values (`0` and `1`) based on the hashed rows. It's possible something a little bit larger would be a bit better, but I haven't tested this much in practice.

**Reply by roelvanderburg** [reply]
Thanks a lot @adrianeboyd, I have a question though:

My [initialize.before_init] does not exists in my config file. Also adding it like below (before or after the [initialize], results in an error).  

```java
[pretaining]

[initialize.before_init]
@callbacks = "customize_tokenizer"

[initialize]
vectors = null
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
lookups = null

[initialize.components]

[initialize.tokenizer]
```

Also the components you shared in your snippet have slightly different versions than mine and have the word senter as pre-fix, but that might just be caused by me initializing the model with the tok2vec component?

```java
[components.senter.model.tok2vec]
@architectures = "spacy.Tok2Vec.v2"

[components.senter.model.tok2vec.embed]
@architectures = "spacy.MultiHashEmbed.v1"
width = ${components.senter.model.tok2vec.encode.width}
attrs = ["NORM","PREFIX","SUFFIX","SHAPE","SPACY"]
rows = [5000,2500,2500,2500,20]
include_static_vectors = false

[components.senter.model.tok2vec.encode]
@architectures = "spacy.MaxoutWindowEncoder.v2"
width = 12
depth = 1
window_size = 1
maxout_pieces = 2

```

**Reply by adrianeboyd** [reply]
If you run `spacy init fill-config`, you should end up with something like this:

```ini
[initialize]
vectors = null
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
lookups = null
before_init = null
after_init = null
```

If you still have an rc version of v3, please upgrade to the most recent release (currently v3.0.1).

You have can either have a pipeline with `tok2vec, senter` where the senter has a tok2vec listener, or the `senter` model has its own internal tok2vec.

The thing to be aware of is that the default `senter` config that you get from `spacy init config` uses a slightly simpler default config, where you can't directly set the attributes, so you need to edit it to use a listener or expand the internal tok2vec config as in the example above to get the more detailed options.

**Reply by roelvanderburg** [reply]
Indeed thanks I did have the rc version. Now I've used 
`spacy init config config.cfg -p senter -l nl -o accuracy -F`
to initialize the config file below:

```ini
[pretraining]

[initialize]
vectors = "nl_core_news_lg"
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
lookups = null
before_init = null
after_init = null

[initialize.components]

[initialize.tokenizer]
```

However it remains quite unclear to me how to refer to my custom callback function in functions.py
https://spacy.io/usage/training#custom-tokenizer
In this page I see the below config as a suggestion:

```ini
[initialize]

[initialize.before_init]
@callbacks = "customize_tokenizer"

```

But inserting those two [initialize.before_init] lines gives me an error:
```
✘ Config validation error
initialize -> before_init   found conflicting values

```

**Reply by adrianeboyd** [reply]
You have both `before_init = null` and the `[initialize.before_init]` block, which are defining the same setting, so you need to pick one.
